File: backend/apps/agents/manager/prompt/compose_turn_system_prompt.py
"""Assemble the full per-turn system prompt for an agent run: the base composed prompt
(default + mode + session) plus the appended context blocks (browser selection, MCP registry
summary, a current-time pin, the App Builder skill, picked app cards, picked Settings rows).
Lifted out of the agent loop; calls the prompt_context builders directly (no manager needed)."""

import logging
import os
from datetime import datetime
from typing import List, Optional

from typeguard import typechecked

from backend.apps.agents.core.models import AgentSession
from backend.apps.agents.manager.prompt.tool_catalog import get_all_tool_names
from backend.apps.agents.manager.prompt.prompt_context import (
    build_app_runtime_contract,
    build_browser_context,
    build_installed_skills_catalog,
    build_mcp_registry_summary,
    build_selected_app_context,
    build_selected_settings_context,
    compose_system_prompt,
)

logger = logging.getLogger(__name__)


@typechecked
def p_resolve_prompt_language() -> str:
    """The UI language the prompt should speak. Unset means pt-BR, matching the frontend default in
    frontend/src/shared/i18n/i18n.ts — a fresh install shows a Portuguese UI, so an English prompt
    there would make the agent answer in the wrong language. Never raises: a turn must not die
    because settings were unreadable."""
    try:
        from backend.apps.settings.store import load_settings
        return "en" if load_settings().language == "en" else "pt-BR"
    except Exception:
        logger.exception("Could not read the language setting; defaulting the prompt to pt-BR")
        return "pt-BR"


@typechecked
def p_localize_default_prompt(default_system_prompt: Optional[str], language: str) -> Optional[str]:
    """Swap in the pt-BR default prompt, but ONLY when the stored value is the untouched English
    default. `default_system_prompt` is a persisted, user-editable setting, so a user who wrote their
    own prompt must get it back verbatim in either language."""
    if language != "pt-BR" or default_system_prompt is None:
        return default_system_prompt
    from backend.apps.settings.models import DEFAULT_SYSTEM_PROMPT, DEFAULT_SYSTEM_PROMPT_PT_BR
    return DEFAULT_SYSTEM_PROMPT_PT_BR if default_system_prompt == DEFAULT_SYSTEM_PROMPT else default_system_prompt


@typechecked
def compose_turn_system_prompt(
    session: AgentSession,
    mode_sys_prompt: Optional[str],
    default_system_prompt: Optional[str],
    selected_browser_ids: Optional[List[str]],
    selected_app_output_ids: Optional[List[str]],
    selected_setting_ids: Optional[List[str]],
) -> Optional[str]:
    # MCP servers and their tool inventories are intentionally NOT injected into the system prompt: the CLI's deferred-tool pool already exposes them by name via ToolSearch, and eagerly listing connected MCPs (account emails, full tool enumerations) here would defeat the deferral and leak every integration into every turn. The model discovers MCPs only when it actively calls ToolSearch; only the gated registry summary goes in.
    browser_ctx = build_browser_context(session.dashboard_id, selected_browser_ids=selected_browser_ids)
    mcp_registry_ctx = build_mcp_registry_summary(session.allowed_tools, session.active_mcps, get_all_tool_names)
    skills_catalog_ctx = build_installed_skills_catalog()
    language = p_resolve_prompt_language()
    composed_prompt = compose_system_prompt(
        p_localize_default_prompt(default_system_prompt, language),
        mode_sys_prompt,
        session.system_prompt,
        browser_ctx,
        mcp_registry_ctx,
        skills_catalog_ctx,
    )

    # Pin the agent's notion of "now" to the host wall clock + zone so it can answer day-of-week questions without hallucinating.
    try:
        from zoneinfo import ZoneInfo
        # Best-effort IANA name for the host. Mirrors apps/service/client.py.
        tz_name = os.environ.get("MAESTRO_TIMEZONE", "").strip()
        if not tz_name:
            try:
                from tzlocal import get_localzone_name  # type: ignore
                tz_name = get_localzone_name() or ""
            except Exception:
                tz_name = ""
        tz_name = tz_name or "UTC"
        now_local = datetime.now(ZoneInfo(tz_name))
        tz_abbr = now_local.strftime("%Z") or tz_name
        time_ctx = (
            "<current_time>\n"
            f"Today is {now_local.strftime('%A, %B %-d, %Y')}.\n"
            f"Local time: {now_local.strftime('%-I:%M %p')} {tz_abbr} ({tz_name}).\n"
            "Use this as ground truth for any date/time/day-of-week question.\n"
            "</current_time>"
        )
        composed_prompt = (composed_prompt + "\n\n" + time_ctx) if composed_prompt else time_ctx
    except Exception:
        pass

    if session.mode == "view-builder":
        # Read the LIVE skill content rather than a frozen-at-import constant. The skill is registered at ~/.claude/skills/app_builder_skill.md; user edits in the Skills page land there and propagate to the agent's prompt next turn without a restart.
        from backend.apps.outputs.view_builder_templates import load_app_builder_skill
        skill_block = f"<app_builder_reference>\n{load_app_builder_skill()}\n</app_builder_reference>"
        composed_prompt = f"{composed_prompt}\n\n{skill_block}" if composed_prompt else skill_block
        # Appended AFTER the reference, and never sourced from it: the skill is a user-editable file seeded once per install, so a stale or edited copy silently drops whatever it omits. Platform mechanics have to reach the agent regardless of what that file says.
        contract_block = build_app_runtime_contract(session.cwd)
        composed_prompt = f"{composed_prompt}\n\n{contract_block}"
    else:
        # Every other mode gets one line of discovery instead of the whole reference: CreateApp's result carries the reference when actually used, so the base prompt stays cheap.
        apps_note = (
            "<apps_capability>\n"
            "You can build real web apps for the user: when they ask for an app, tool, game, or dashboard, "
            "call the CreateApp tool — it seeds a workspace and puts a live preview card on their dashboard, "
            "then you write the code. To change an existing app, have the user select its card (or use the "
            "workspace path in your context) and edit the files directly.\n"
            "</apps_capability>"
        )
        composed_prompt = f"{composed_prompt}\n\n{apps_note}" if composed_prompt else apps_note

    # App cards the user picked via the dashboard element picker: give the agent each app's on-disk path + meta + SKILL.md pointer so it can edit them in place (the dashboard card's runtime live-reloads). Additive and independent of view-builder mode above.
    app_ctx = build_selected_app_context(selected_app_output_ids)
    if app_ctx:
        composed_prompt = f"{composed_prompt}\n\n{app_ctx}" if composed_prompt else app_ctx

    # The user can point the agent at specific Settings rows. Targeting aid only; the settings tools are always on regardless.
    settings_ctx = build_selected_settings_context(selected_setting_ids)
    if settings_ctx:
        composed_prompt = f"{composed_prompt}\n\n{settings_ctx}" if composed_prompt else settings_ctx

    # Last block in the prompt, so it wins over any English phrasing in the blocks above it.
    if language == "pt-BR":
        lang_ctx = (
            "<language_directive>\n"
            "Escreva em português do Brasil todo o conteúdo que o usuário vai ler: respostas em prosa, resumos, "
            "planos, títulos de tarefas e de itens de TODO, perguntas do AskUserQuestion e mensagens de erro.\n"
            "Mantenha exatamente como estão, sem traduzir: código e identificadores, caminhos de arquivo, comandos "
            "de shell, saída de log, nomes de ferramentas e conteúdo citado de arquivos ou da web.\n"
            "Não traduza nomes de produtos (Maestro Studio, Claude, Anthropic, MCP, GitHub) nem os termos técnicos "
            "que pessoas da área usam em inglês no Brasil (prompt, token, workflow, deploy, commit, build, log).\n"
            "</language_directive>"
        )
    else:
        lang_ctx = (
            "<language_directive>\n"
            "Write all user-facing content in English: prose replies, summaries, plans, task and TODO titles, "
            "AskUserQuestion prompts, and error messages.\n"
            "Leave code, identifiers, file paths, shell commands, log output, and quoted content as-is.\n"
            "</language_directive>"
        )
    composed_prompt = f"{composed_prompt}\n\n{lang_ctx}" if composed_prompt else lang_ctx

    return composed_prompt

File: engine/src/agents/manager/prompt/composeTurnSystemPrompt.ts
// engine/src/agents/manager/prompt/composeTurnSystemPrompt.ts -- AGT-5, a port of
// backend/apps/agents/manager/prompt/compose_turn_system_prompt.py: assembles the full per-turn
// system prompt (base composed prompt + appended context blocks: browser selection, MCP registry
// summary, a current-time pin, the App Builder skill, picked app cards, picked Settings rows, the
// language directive).
//
// Every context-builder call is DI'd (`ComposeTurnSystemPromptDeps`), defaulting to promptContext.ts's
// exports -- exactly the functions backend/tests/test_system_prompt.py patches
// (`build_browser_context`/`build_mcp_registry_summary`/`build_selected_app_context`/
// `build_selected_settings_context`), plus `loadSettings` (language resolution) and
// `loadAppBuilderSkill` (the view-builder skill body, outputs.view_builder_templates -- SUB-5
// territory, not yet ported; default returns a placeholder documenting the gap rather than
// silently omitting the block the real prompt always carries in view-builder mode).

import type { AgentSession } from '../../core/models';
import { loadSettings } from '../../../settings/store';
import { DEFAULT_SYSTEM_PROMPT, DEFAULT_SYSTEM_PROMPT_PT_BR } from '../../../settings/models';
import {
  buildAppRuntimeContract,
  buildBrowserContext as defaultBuildBrowserContext,
  buildInstalledSkillsCatalog as defaultBuildInstalledSkillsCatalog,
  buildMcpRegistrySummary as defaultBuildMcpRegistrySummary,
  buildSelectedAppContext as defaultBuildSelectedAppContext,
  buildSelectedSettingsContext as defaultBuildSelectedSettingsContext,
  composeSystemPrompt,
} from './promptContext';
import { getAllToolNames as realGetAllToolNames } from '../../../apps/toolsLib/toolCatalog';

/** Full call-through to tool_catalog.py's `get_all_tool_names` port (SUB-4). */
function defaultGetAllToolNames(): string[] {
  return realGetAllToolNames();
}

/** Stand-in for `outputs.view_builder_templates.load_app_builder_skill` -- outputs (SUB-5, App
 * Builder) isn't ported. Returns a placeholder rather than silently omitting the block the real
 * prompt always carries in view-builder mode, so a missing wiring is visible in the prompt itself
 * rather than a quiet behavior gap. */
function defaultLoadAppBuilderSkill(): string {
  return '(App Builder skill content not yet available -- outputs/view_builder_templates is not ported to the engine.)';
}

export interface ComposeTurnSystemPromptDeps {
  buildBrowserContext?: typeof defaultBuildBrowserContext;
  buildMcpRegistrySummary?: typeof defaultBuildMcpRegistrySummary;
  buildSelectedAppContext?: typeof defaultBuildSelectedAppContext;
  buildSelectedSettingsContext?: typeof defaultBuildSelectedSettingsContext;
  buildInstalledSkillsCatalog?: typeof defaultBuildInstalledSkillsCatalog;
  getAllToolNames?: () => string[];
  loadAppBuilderSkill?: () => string;
  /** Mirrors `p_resolve_prompt_language`'s `load_settings` call -- defaults to the real (never-
   * throwing) engine settings loader; kept overridable so a test can simulate the Python
   * original's OSError fail-open path without touching disk. */
  loadSettings?: () => { settings: { language: string | null } };
  /** Wall-clock source for the `<current_time>` pin, overridable for determinism. */
  now?: () => Date;
  timeZone?: string;
}

/** The UI language the prompt should speak. Unset means pt-BR, matching the frontend default -- a
 * fresh install shows a Portuguese UI, so an English prompt there would make the agent answer in
 * the wrong language. Never raises: a turn must not die because settings were unreadable. */
export function resolvePromptLanguage(loadSettingsFn: () => { settings: { language: string | null } } = () => loadSettings()): string {
  try {
    return loadSettingsFn().settings.language === 'en' ? 'en' : 'pt-BR';
  } catch {
    return 'pt-BR';
  }
}

/** Swap in the pt-BR default prompt, but ONLY when the stored value is the untouched English
 * default. `defaultSystemPrompt` is a persisted, user-editable setting, so a user who wrote their
 * own prompt must get it back verbatim in either language. */
export function localizeDefaultPrompt(defaultSystemPrompt: string | null | undefined, language: string): string | null | undefined {
  if (language !== 'pt-BR' || defaultSystemPrompt === null || defaultSystemPrompt === undefined) return defaultSystemPrompt;
  return defaultSystemPrompt === DEFAULT_SYSTEM_PROMPT ? DEFAULT_SYSTEM_PROMPT_PT_BR : defaultSystemPrompt;
}

function formatCurrentTimeBlock(now: Date, timeZone: string): string {
  const dateFmt = new Intl.DateTimeFormat('en-US', { timeZone, weekday: 'long', month: 'long', day: 'numeric', year: 'numeric' });
  const timeFmt = new Intl.DateTimeFormat('en-US', { timeZone, hour: 'numeric', minute: '2-digit', hour12: true, timeZoneName: 'short' });
  const dateStr = dateFmt.format(now);
  const timeParts = timeFmt.formatToParts(now);
  const clock = timeParts
    .filter((p) => p.type === 'hour' || p.type === 'minute' || p.type === 'dayPeriod')
    .map((p) => p.value)
    .join('')
    .replace(/([AP]M)$/, ' $1');
  const tzAbbr = timeParts.find((p) => p.type === 'timeZoneName')?.value ?? timeZone;
  return (
    '<current_time>\n' +
    `Today is ${dateStr}.\n` +
    `Local time: ${clock} ${tzAbbr} (${timeZone}).\n` +
    'Use this as ground truth for any date/time/day-of-week question.\n' +
    '</current_time>'
  );
}

export interface ComposeTurnSystemPromptArgs {
  modeSysPrompt: string | null;
  defaultSystemPrompt: string | null;
  selectedBrowserIds: string[] | null;
  selectedAppOutputIds: string[] | null;
  selectedSettingIds: string[] | null;
}

/** Full port of `compose_turn_system_prompt`. */
export function composeTurnSystemPrompt(
  session: Pick<AgentSession, 'dashboard_id' | 'allowed_tools' | 'active_mcps' | 'system_prompt' | 'mode' | 'cwd'>,
  args: ComposeTurnSystemPromptArgs,
  deps: ComposeTurnSystemPromptDeps = {},
): string | undefined {
  const buildBrowserContext = deps.buildBrowserContext ?? defaultBuildBrowserContext;
  const buildMcpRegistrySummary = deps.buildMcpRegistrySummary ?? defaultBuildMcpRegistrySummary;
  const buildSelectedAppContext = deps.buildSelectedAppContext ?? defaultBuildSelectedAppContext;
  const buildSelectedSettingsContext = deps.buildSelectedSettingsContext ?? defaultBuildSelectedSettingsContext;
  const buildInstalledSkillsCatalog = deps.buildInstalledSkillsCatalog ?? defaultBuildInstalledSkillsCatalog;
  const getAllToolNames = deps.getAllToolNames ?? defaultGetAllToolNames;
  const loadAppBuilderSkill = deps.loadAppBuilderSkill ?? defaultLoadAppBuilderSkill;
  const now = (deps.now ?? (() => new Date()))();
  const timeZone =
    deps.timeZone ?? (process.env.MAESTRO_TIMEZONE?.trim() || Intl.DateTimeFormat().resolvedOptions().timeZone || 'UTC');

  // MCP servers and their tool inventories are intentionally NOT injected into the system prompt
  // eagerly -- see the Python original's comment for the full reasoning (ToolSearch deferral).
  const browserCtx = buildBrowserContext(session.dashboard_id, args.selectedBrowserIds);
  const mcpRegistryCtx = buildMcpRegistrySummary(session.allowed_tools, session.active_mcps, getAllToolNames);
  const skillsCatalogCtx = buildInstalledSkillsCatalog();
  const language = resolvePromptLanguage(deps.loadSettings);
  let composed: string | undefined = composeSystemPrompt(
    localizeDefaultPrompt(args.defaultSystemPrompt, language),
    args.modeSysPrompt,
    session.system_prompt,
    browserCtx,
    mcpRegistryCtx,
    skillsCatalogCtx,
  );

  // Pin the agent's notion of "now" to the host wall clock + zone.
  try {
    const timeCtx = formatCurrentTimeBlock(now, timeZone);
    composed = composed ? `${composed}\n\n${timeCtx}` : timeCtx;
  } catch {
    // Best-effort, mirrors the Python original's bare except: pass.
  }

  if (session.mode === 'view-builder') {
    // Read the LIVE skill content rather than a frozen-at-import constant.
    const skillBlock = `<app_builder_reference>\n${loadAppBuilderSkill()}\n</app_builder_reference>`;
    composed = composed ? `${composed}\n\n${skillBlock}` : skillBlock;
    // Appended AFTER the reference, and never sourced from it: platform mechanics have to reach
    // the agent regardless of what that (user-editable) file says.
    const contractBlock = buildAppRuntimeContract(session.cwd);
    composed = `${composed}\n\n${contractBlock}`;
  } else {
    // Every other mode gets one line of discovery instead of the whole reference.
    const appsNote =
      '<apps_capability>\n' +
      'You can build real web apps for the user: when they ask for an app, tool, game, or dashboard, ' +
      'call the CreateApp tool — it seeds a workspace and puts a live preview card on their dashboard, ' +
      'then you write the code. To change an existing app, have the user select its card (or use the ' +
      'workspace path in your context) and edit the files directly.\n' +
      '</apps_capability>';
    composed = composed ? `${composed}\n\n${appsNote}` : appsNote;
  }

  // App cards the user picked via the dashboard element picker.
  const appCtx = buildSelectedAppContext(args.selectedAppOutputIds);
  if (appCtx) composed = composed ? `${composed}\n\n${appCtx}` : appCtx;

  // The user can point the agent at specific Settings rows.
  const settingsCtx = buildSelectedSettingsContext(args.selectedSettingIds);
  if (settingsCtx) composed = composed ? `${composed}\n\n${settingsCtx}` : settingsCtx;

  // Last block in the prompt, so it wins over any English phrasing in the blocks above it.
  const langCtx =
    language === 'pt-BR'
      ? '<language_directive>\n' +
        'Escreva em português do Brasil todo o conteúdo que o usuário vai ler: respostas em prosa, resumos, ' +
        'planos, títulos de tarefas e de itens de TODO, perguntas do AskUserQuestion e mensagens de erro.\n' +
        'Mantenha exatamente como estão, sem traduzir: código e identificadores, caminhos de arquivo, comandos ' +
        'de shell, saída de log, nomes de ferramentas e conteúdo citado de arquivos ou da web.\n' +
        'Não traduza nomes de produtos (Maestro Studio, Claude, Anthropic, MCP, GitHub) nem os termos técnicos ' +
        'que pessoas da área usam em inglês no Brasil (prompt, token, workflow, deploy, commit, build, log).\n' +
        '</language_directive>'
      : '<language_directive>\n' +
        'Write all user-facing content in English: prose replies, summaries, plans, task and TODO titles, ' +
        'AskUserQuestion prompts, and error messages.\n' +
        'Leave code, identifiers, file paths, shell commands, log output, and quoted content as-is.\n' +
        '</language_directive>';
  composed = composed ? `${composed}\n\n${langCtx}` : langCtx;

  return composed;
}
